File: trunk/src/Animation/EvalTimeflow.cpp
#include <Bibim/Config.h>
#include <Bibim/EvalTimeflow.h>
#include <Bibim/ComponentStreamReader.h>
#include <Bibim/Math.h>
#include <Bibim/Timeline.h>

namespace Bibim
{
    BBImplementsComponent(EvalTimeflow);

    EvalTimeflow::EvalTimeflow()
        : evaluatedTime(0.0f),
          timeline(nullptr),
          time(0.0f),
          duration(0.0f),
          velocity(1.0f),
          stopBehavior(PauseAtStop),
          looped(true),
          normalized(false),
          isUpdating(false)
    {
    }

    EvalTimeflow::~EvalTimeflow()
    {
        if (timeline != nullptr && isUpdating)
            timeline->Remove(this);
    }

    void EvalTimeflow::Update(float dt, int timestamp)
    {
        if (lastTimestamp == timestamp || velocity == 0.0f)
            return;
        else
            lastTimestamp = timestamp;

        const float nextTime = time + (dt * velocity);

        if (0 <= nextTime && nextTime <= duration)
            time = nextTime;
        else
        {
            if (nextTime > duration)
            {
                if (looped && duration > 0.0f)
                    time = Math::Mod(nextTime, duration);
                else
                    time = duration;
            }
            else
            {
                if (looped && duration > 0.0f)
                    time = duration - Math::Mod(-nextTime, duration);
                else
                    time = 0.0f;
            }
        }

        if (normalized)
        {
            if (duration > 0.0f)
                evaluatedTime = time / duration;
            else
                evaluatedTime = 0.0f;
        }
        else
            evaluatedTime = time;
    }

    void EvalTimeflow::Start()
    {
        if (isUpdating)
            return;

        isUpdating = true;
        if (timeline != nullptr)
            timeline->Add(this, this);
    }

    void EvalTimeflow::Stop()
    {
        if (isUpdating == false)
            return;

        isUpdating = false;
        if (timeline != nullptr)
            timeline->Remove(this);

        switch (stopBehavior)
        {
            case PauseAtStop:
                break;
            case ResetAtStop:
                time = 0.0f;
                evaluatedTime = 0.0f;
                break;
            case FinishAtStop:
                time = duration;
                if (normalized)
                    evaluatedTime = 1.0f;
                else
                    evaluatedTime = duration;
                break;
        }
    }

    void EvalTimeflow::Reset()
    {
        time = 0.0f;
        evaluatedTime = 0.0f;
    }

    float EvalTimeflow::Evaluate(EvalContext& /*context*/)
    {
        return evaluatedTime;
    }

    void EvalTimeflow::SetTimeline(Timeline* value)
    {
        if (timeline != value)
        {
            if (timeline != nullptr && isUpdating)
                timeline->Remove(this);

            timeline = value;

            if (timeline != nullptr && isUpdating)
                timeline->Add(this);
        }
    }

    void EvalTimeflow::SetDuration(float value)
    {
        value = Math::Max(value, 0.0f);
    }

    void EvalTimeflow::OnRead(ComponentStreamReader& reader)
    {
        Base::OnRead(reader);
        evaluatedTime = 0.0f;
        timeline = static_cast<Timeline*>(reader.ReadModule(Timeline::ClassID));
        time = 0.0f;
        duration = reader.ReadFloat();
        velocity = reader.ReadFloat();
        stopBehavior = static_cast<Behavior>(reader.ReadByte());
        looped = reader.ReadBool();
        normalized = reader.ReadBool();
    }

    void EvalTimeflow::OnCopy(const GameComponent* original, CloningContext& context)
    {
        Base::OnCopy(original, context);
        const This* o = static_cast<const This*>(original);
        evaluatedTime = o->evaluatedTime;
        timeline = o->timeline;
        time = o->time;
        duration = o->duration;
        velocity = o->velocity;
        stopBehavior = o->stopBehavior;
        looped = o->looped;
        normalized = o->normalized;
        isUpdating = o->isUpdating;

        if (timeline != nullptr && isUpdating)
            timeline->Add(this);
    }
}

File: tools/Bibim.Toolkit/Animation/Evaluation/EvalTimeflow.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using Bibim.Time;

namespace Bibim.Animation
{
    [ClassID('e', 'T', 'F', 'f')]
    public sealed class EvalTimeflow : Eval<float>, IUpdateable
    {
        #region Behavior (Nested Enum)
        public enum Behavior
        {
            Pause,
            Reset,
            Finish,
        }
        #endregion

        #region Fields
        private float evaluatedTime;
        private Timeline timeline;
        private float duration;
        private int lastTimestamp;
        private bool isUpdating;
        #endregion

        #region Properties
        public Timeline Timeline
        {
            get { return timeline; }
            set
            {
                if (timeline != value)
                {
                    if (timeline != null && isUpdating)
                        timeline.Remove(this);

                    timeline = value;

                    if (timeline != null && isUpdating)
                        timeline.Add(this);
                }
            }
        }

        public float Time
        {
            get;
            private set;
        }

        public float Duration
        {
            get { return duration; }
            set
            {
                duration = Math.Max(value, 0.0f);
            }
        }

        public float Velocity
        {
            get;
            set;
        }

        public bool IsLooped
        {
            get;
            set;
        }

        public bool IsNormalized
        {
            get;
            set;
        }

        public Behavior StopBehavior
        {
            get;
            set;
        }
        #endregion

        #region Constructors
        public EvalTimeflow()
            : this(0.0f, 1.0f, Behavior.Pause, true, false)
        {
        }

        public EvalTimeflow(float duration)
            : this(duration, 1.0f, Behavior.Pause, true, false)
        {
        }

        public EvalTimeflow(float duration, float velocity, Behavior stopBehavior, bool isLooped, bool isNormalized)
        {
            Time = 0.0f;
            Duration = duration;
            Velocity = velocity;
            StopBehavior = stopBehavior;
            IsLooped = isLooped;
            IsNormalized = isNormalized;
        }
        #endregion

        #region Methods
        public void Update(float dt, int timestamp)
        {
            if (lastTimestamp == timestamp || Velocity == 0.0f)
                return;
            else
                lastTimestamp = timestamp;

            float nextTime = Time + dt;

            if (0 <= nextTime && nextTime <= duration)
                Time = nextTime;
            else
            {
                if (nextTime < duration)
                {
                    if (IsLooped && duration > 0.0f)
                        Time = nextTime % duration;
                    else
                        Time = duration;
                }
                else
                {
                    if (IsLooped && duration > 0.0f)
                        Time = duration - (-nextTime % duration);
                    else
                        Time = 0.0f;
                }
            }

            if (IsNormalized)
            {
                if (Duration > 0.0f)
                    evaluatedTime = Time / Duration;
                else
                    evaluatedTime = 0.0f;
            }
            else
                evaluatedTime = Time;
        }
        #endregion

        public override void Start()
        {
            if (isUpdating)
                return;

            isUpdating = true;
            if (timeline != null)
                timeline.Add(this);
        }

        public override void Stop()
        {
            if (isUpdating == false)
                return;

            isUpdating = false;
            if (timeline != null)
                timeline.Remove(this);

            switch (StopBehavior)
            {
                case Behavior.Pause:
                    break;
                case Behavior.Reset:
                    Time = 0.0f;
                    break;
                case Behavior.Finish:
                    Time = Duration;
                    break;
            }
        }

        public override void Reset()
        {
            Time = 0.0f;
            evaluatedTime = 0.0f;
        }

        public override float Evaluate(EvalContext context)
        {
            return evaluatedTime;
        }
    }
}
